### PR TITLE
[AST/Sema] Distributed: Refactor type and member queries

### PR DESCRIPTION
**Pull request by xedin**
- Simplifies existing associated type handling logic 
- Introduces a unified way to retrieve serialization requirement for distributed actors
- Moves some of the APIs out of ASTContext
- Addresses some of the FIXMEs.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

File: lib/AST/DistributedDecl.cpp [hunk 1]
```diff
@@ -163,68 +127,96 @@ Type swift::getDistributedActorSystemType(NominalTypeDecl *actor) {
...
+    SmallVector<Type, 2> members;
+    llvm::transform(protocols, std::back_inserter(members), [](const auto *P) {
+      return P->getDeclaredInterfaceType();
+    });
+
+    return ExistentialType::get(
+        ProtocolCompositionType::get(ctx, members,
+                                     /*inverses=*/{},
+                                     /*HasExplicitAnyObject=*/false));
+  }
```
Comment (ktoso) [inline lib/AST/DistributedDecl.cpp:180]: Nice

**Review by ktoso — APPROVED** [review]
Great cleanup thank you!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

File: lib/AST/DistributedDecl.cpp [hunk 2]
```diff
@@ -163,68 +127,96 @@ Type swift::getDistributedActorSystemType(NominalTypeDecl *actor) {
...
+
+    auto actorProtocol = ctx.getProtocol(KnownProtocolKind::DistributedActor);
+    if (!actorProtocol)
+      return Type();
+
+    auto serializationTy =
+        actorProtocol->getAssociatedType(ctx.Id_SerializationRequirement)
+            ->getDeclaredInterfaceType();
+
+    auto protocols = sig->getRequiredProtocols(serializationTy);
```
Comment (slavapestov) [inline lib/AST/DistributedDecl.cpp:167]: This whole thing is just `sig->getExistentialType(serializationTy)`

Comment (xedin) [reply lib/AST/DistributedDecl.cpp:167]: I got spooked by the comment there, will give it a try!

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/AST/DistributedDecl.cpp (hunk 2 above)
Comment (xedin) [reply lib/AST/DistributedDecl.cpp]: Fixed it, thank you!

**Pull request merged**